Adding blog post about the making of tutorials (#172)
temporalio/documentation: blog/2020-10-23-the-making-of-new-user-sdk-tutorials.md

At a high level, there are two parts to Temporal: the backend and a client-side SDK. Configuring, deploying, and operating the Temporal backend for a live environment is no small task. On the other hand, it is really easy to [get Temporal running on your local machine](/docs/install-temporal-server) in a Docker container. In fact, you can do it with just two terminal commands.

The Docker route definitely simplifies running the backend, which means the majority of friction for new users comes from our SDKs ([Go](https://github.com/temporalio/sdk-go), [Java](https://github.com/temporalio/sdk-java)). While an SDK is meant to abstract the complexities of interacting with the server API, Temporal flips a lot of the preconceived notions of modern application development on their head. The SDK docs needed to do more than just provide example usage. They also needed to show why, and allow the user to grasp the concepts that Temporal is promoting. So we went about scoping something that we could realistically accomplish within that time frame and still be relatively effective.

> We decided that the best thing for us to focus on was a great first experience for new users. We wanted something that would enable a developer to start using the product right away but also instill an understanding of the value it provides. We wanted to cultivate that "aha!" moment.

Author: 👍

We started as most might, by trying to envision what the ideal new user experience would look like. We then identified as many of the steps it would take to get there as possible. Looking back, I would contend that we managed to lock onto three ideas that we thought would get us closer to the ideal experience. The hope was that once these three ideas were combined they would result in a set of effective SDK tutorials for new users.

## Snipsync

<img alt="image" class="docs-image-centered docs-image-max-width-20" src="https://raw.githubusercontent.com/temporalio/documentation-images/main/static/sync.png" />

> It was around this time (early September), that I was testing out a NodeJS tool I had built to improve the experience of creating and maintaining documentation. It downloads Github repos, scrapes code snippets that exist between specific comment wrappers, and writes the snippets to their corresponding comment wrappers in Markdown files. We borrowed the idea from [Google's proprietary version](https://github.com/GoogleCloudPlatform/golang-samples/blob/master/secretmanager/get_secret.go#L17) they use for their [Google Cloud documentation](https://cloud.google.com/docs). The concept is fairly simple, and I am still surprised that there was no existing open-source solution. So we made one!

Reviewer: I think this section is far better but "commit wrapper" is not a familiar term (at least to me).

Reviewer: I think showing a very concise example would go a long way.

A tool that automates the synchronization of code with the docs from any given repository has several key benefits:

1. Code snippets in the documentation are guaranteed to work as they are continuously tested. This also means that they can be reliably copied and pasted into the user's editor.
2. We control exactly which lines of code are shown and can also target a specific branch or commit. This is a great safeguard against bugs which might be introduced to the main branch.
3. We never have to commit source code into the docs. The code is merged into the Markdown at build time. This ensures that the code is already reviewed and approved from the repo it resides in.

Snipsync does come with a few considerations:

1. Embedded code needs to have carefully reviewed comments, structure, and make sense within the context of the documentation. For example, if the code snippet is coming from a working repo it may include additional variables or function calls. These must be minimized and optimized out so they don’t cause unnecessary confusion.

> 2. In the same way that the code must be optimized for the docs, the docs must also be optimized for the code. In essence, the docs are being “driven” and "defined" by the underlying code. And if no one has coined the term yet, I think the credit for “code-driven documentation” should go to our Head of Product, Ryland Goldstein as he pinged me early one morning to share that epiphany with me.

Reviewer: ![Screen Shot 2020-11-04 at 3 03 58 PM](https://user-images.githubusercontent.com/27736122/98177270-1bd53600-1eaf-11eb-97f3-60b76267a3c7.png)

You can use this!

> We decided to embrace [Snipsync](https://github.com/temporalio/snipsync) and solve the challenges that it introduced.

Reviewer: ```suggestion
We decided to embrace [Snipsync](https://github.com/temporalio/snipsync) as the challenges it introduces are minimal compared to the value.
```

## Template repos

<img alt="image" class="docs-image-centered docs-image-max-width-20" src="https://raw.githubusercontent.com/temporalio/documentation-images/main/static/templates.png" />

We now had a way to synchronize code with our documentation. But from where will the code be synchronized? We know users will likely want to view the source file and relative file path of the code snippet for added context. They will also be likely to clone the repo and try to run the sample.

> We actually already repositories of code samples for the [Go SDK](https://github.com/temporalio/samples-go) and [Java SDK](https://github.com/temporalio/samples-java). And while we desired to see more samples, there were already quite a few of them in each. But, we discovered that shared sample repositories tended to have two issues that made them less ideal for synchronizing with docs.

Reviewer: Typo in the first sentence, but also I really don't recommend starting a new para with "actually'. It makes the reader feel out of the loop.

Reviewer: ```suggestion
We actually already repositories of code samples for the [Go SDK](https://github.com/temporalio/samples-go) and [Java SDK](https://github.com/temporalio/samples-java). While we desired to see more samples, there were already quite a few of them in each repository. But, we discovered that shared sample repositories tend to have two issues that made them less ideal for synchronizing with docs.
```

1. While convenient, colocating multiple samples in a single repo is far less approachable compared with storing samples in self-contained repos.
2. In shared sample repositories, it is harder to retain the idiomatic nature of any language and mirror the functionality of any sample across different languages.

So for each of the sample applications we planned on using to drive our documentation, we created a corresponding template repo. These template repos can be easily copied, built, and run in a matter of minutes.

- [Go money transfer template repo](https://github.com/temporalio/money-transfer-project-template-go)
- [Go hello world template repo](https://github.com/temporalio/hello-world-project-template-go)
- [Java money transfer template repo](https://github.com/temporalio/money-transfer-project-template-java)
- [Java hello world template repo](https://github.com/temporalio/hello-world-project-template-java)

## The tutorial

<img alt="image" class="docs-image-centered docs-image-max-width-20" src="https://raw.githubusercontent.com/temporalio/documentation-images/main/static/tutorials.png" />

> Since the goal of our documentation changes was to help with new user acquisition, we decided to aim for a "tutorial" style of documentation. The first iterations aimed to build upon and replace the existing SDK “quick start” pages that maintained the status quo and printed “Hello World!” to the console. As you may have guessed, this route wasn’t sufficient enough to show users the real values that Temporal offers.

Reviewer: ```suggestion
Since the goal of our documentation changes was to help with new user acquisition, we decided to aim for a "tutorial" style of documentation. The first iterations aimed to build upon and replace the existing SDK “quick start” pages that maintained the status quo and printed “Hello World!” to the console. As you may have guessed, this route wasn’t sufficient enough to show users the real value Temporal offers.
```

> Once it became clear that a standard approach wasn't going to cut it, we approached our co-founder and CEO, Maxim Fateev. We asked him to give us a demonstration that he typically uses to introduce engineers to Temporal for the first time. The scenario represents a money transfer from one bank account to another and he demonstrates what happens if one of the steps in the transfer fails. The money transfer sample was a great way to introduce the values of Temporal. For if you understand the ramifications of losing money from a failed financial transaction, several values of Temporal become immediately apparent:

Reviewer: ```suggestion
Once it became clear that a standard approach wasn't going to cut it, we brought in our co-founder and CEO, Maxim Fateev. We asked him to give us a demonstration that he typically uses to introduce engineers to Temporal for the first time. The scenario represents a money transfer from one bank account to another and during the demo Maxim showcases what happens if one of the steps in the transfer fails. The money transfer sample was a great way to introduce the values of Temporal. For if you understand the ramifications of losing money from a failed financial transaction, several values of Temporal become immediately apparent:
```

Reviewer: So I think this sentence is indicative of a minimal but widespread issue. 

"For if you understand the ramifications of losing money from a failed financial transaction, several values of Temporal become immediately apparent:"

It's such a great message but its just wayyy too wordy. I feel like this is borderline shakespearian. I think this is true in quite a few other places. Anyway, just try and keep an eye on that.

Author: But this is also a Blog post - not docs.
I think that blog posts should not have to wait for the slowest reader - creative, though accurate, wording should be encouraged.

1. The state of running code is maintained even through hardware failures, server crashes, and network outages.
2. There is deep visibility into the state of code execution out of the box via the CLI or UI.
3. Function calls come with automatic and retries and configurable timeouts.
4. Bugs can be hot fixed in running code.

> For someone that is new to Temporal the attraction doesn’t come from using the SDK to print “Hello World!”, but instead comes from witnessing the inherent benefits that Temporal offers by simply running simulations using a pre-built application.

Reviewer: Should be two sentences

This is the direction that we decided to send new users. If a user can wrap their heads around the value that Temporal brings to their application right out of the box, then spending time and energy on application setup and learning the SDK becomes a non-blocker.

Check out these finished tutorials and see for yourself!

- [Go: Run your first Temporal application](/docs/go-run-your-first-app)
- [Java: Run your first Temporal application](/docs/java-run-your-first-app)

## Next steps

<img alt="image" class="docs-image-centered docs-image-max-width-20" src="https://raw.githubusercontent.com/temporalio/documentation-images/main/static/steps.png" />

> At Temporal we understand that our documentation plays a very important role in our users' experience. And to get our docs into a world class state we have lots of work in front of us. In our next steps we will be looking at the end-to-end journey through our documentation and how we can provide the best experience for every user. To validate any direction we take we will be engaging with the community to hear what you think and help us dial things in. Any user can [schedule a 15 minute feedback session](https://calendly.com/cully-temporal/schedule-feedback-session) with me directly! We will also be preparing for all the new and exciting features around our hosted cloud offering that will enable all developers to build invincible applications.

Reviewer: "In our next steps" awkward phrasing IMO

"To validate any direction we take we will be engaging with the com" - comma

id: the-making-of-new-user-sdk-tutorials
title: The making of the new user SDK tutorials
author: Cully Wakelin
author_title: SDE Technical Writer
author_image_url: https://avatars2.githubusercontent.com/u/34380806?s=400&u=5cd38b5e4416a5d10cdf9ebd386eec1d02f0b067&v=4
image: /img/docs/sketchbook.png
release_version: V1.1.1
---

<img class="docs-image-centered docs-image-max-width-20" src={require('../static/img/docs/sketchbook.png').default} />

<!--truncate-->

&nbsp;&nbsp; Around the beginning of September it became clear that v1.0 of Temporal would be released by the end of the month and we would be ready to announce our funding shortly thereafter. So, from the perspective of our documentation, we wanted to align and prioritize our changes such that they amplified the effects of these events. With any product launch, we were hoping to see a buzz around the product and a surge in new users. And in the case of the latter, documentation would play an important role in introducing new users to the product and orienting them on their journey.

## Uphill challenges

<img class="docs-image-centered docs-image-max-width-20" src={require('../static/img/docs/hiker.png').default} />

> &nbsp;&nbsp; In terms of our documentation, it’s no secret that we were facing a number of challenges. To start, product documentation is typically offered a secondary level of prioritization; Faced with limited resources, as long as the product ships with some minimal viable information and access to product experts who can answer questions, then at least most advanced users can get it going. So, documentation tends to lag behind in fully describing a product’s capability and feature set, which held true for Temporal.

Author: Ya, understood rewriting.

But, this also introduces a couple of challenges:

1. The code needs to have carefully reviewed comments, structure, and make sense within the context of the documentation. For example, if the code snippet is coming from a working repo it may include additional variables or function calls which must be minimized and optimized so that they don’t cause unnecessary confusion.
2. In the same way that the code must be optimized for the docs, the docs must be optimized for the code. In essence, the docs are being “driven” by the code. And if no one has coined the term yet, I think the credit for “code-driven documentation” should go to our Head of Product, Ryland Goldstein for he pinged me early one morning to share that epiphany with me.
We decided to try [Snipsync](https://github.com/temporalio/snipsync) and solve the challenges that it introduced.

## Template repos

<img class="docs-image-centered docs-image-max-width-20" src={require('../static/img/docs/templates.png').default} />

> &nbsp;&nbsp; We actually had a repository of code samples for the Go SDK and Java SDK already. And while we desired to see more samples, there were already quite a few of them in each. However we discovered that users will likely want to view the source file and relative file path of the code snippet for added context. They will also be likely to clone the repo and try to run the sample. Shared sample repositories as we discovered tended to have two issues:

Author: Thanks, ya this needed more setup and context.

1. While effective, they have more overhead in terms of approachability than that of stand alone repositories that have a single fully working application sample designed to illustrate certain concepts.
2. When it comes to SDKs in multiple languages and shared repos, it is harder to retain the idiomatic nature of the languages and mirror the functionality of the samples across the different languages than if each sample existed in its own repo.

So, we created template repos for each of the sample applications that we planned on using in our docs. The templates can be easily copied, built, and run in a matter of minutes.

- [Go money transfer template repo](https://github.com/temporalio/money-transfer-project-template-go)
- [Go hello world template repo](https://github.com/temporalio/hello-world-project-template-go)
- [Java money transfer template repo](https://github.com/temporalio/money-transfer-project-template-java)
- [Java hello world template repo](https://github.com/temporalio/hello-world-project-template-java)

## The tutorial

<img class="docs-image-centered docs-image-max-width-20" src={require('../static/img/docs/tutorials.png').default} />

> &nbsp;&nbsp; Since we really wanted new users to grasp some of the core value propositions Temporal offers, we decided to aim for a “tutorial” style of documentation. The first iterations aimed to build upon and replace the existing SDK “quick start” pages that maintained the industry standard of showing an example that printed “Hello World!” to the console. But, as you may have guessed, this route wasn’t sufficient enough to show users the real values that Temporal offers.

Author: 👍

> So, we approached our co-founder and CEO, Maxim Fateev, and asked him to give us the same demonstration that he liked to give when introducing the software to another engineer for the first time. One of Maxim’s go-to scenarios is a money transfer from one bank account to another. “Aha!” we said, because as it turns out, for someone new to Temporal the attraction doesn’t come from using the SDK to print “Hello World!”, but it is instead to witness the inherent benefits that Temporal offers by running simulations using a pre-built application.

Author: Restructured this whole part.

Because everyone can understand the ramifications of losing money during a transaction, the money transfer demonstration shows off at least four of Temporal’s main value propositions:

1. The state of the code execution is maintained even through server outages and crashes.
2. There is deep visibility into the state of the code execution via the CLI or UI.

> 3. Code has automatic timeouts and retries, which are easily configurable.

Reviewer: ```suggestion
3. Function calls come with automatic timeouts and retries, which are easily configurable.
```

I don't think timeouts are automatic

> 4. Bugs can be fixed “on-the-fly” without losing the state of the code execution.

Reviewer: ```suggestion
4. Bugs can be hot fixed in running code
```

> So, it was this direction that we decided to send new users. Once a user can wrap their heads around the value that Temporal brings to their application right out of the box, then spending time and energy on application setup and learning the SDK becomes trivial.

Reviewer: ```suggestion
This is the direction that we decided to send new users. If a user can wrap their heads around the value that Temporal brings to their application out of the box, spending time and energy on application setup and learning the SDK becomes a non-blocker.
```

## Next steps

<img class="docs-image-centered docs-image-max-width-20" src={require('../static/img/docs/steps.png').default} />

> &nbsp;&nbsp; We are far from satisfied with our documentation. We plan to bring SDK reference right into the docs, merge SDK tutorials into a single journey, so that a user can easily see how things are done across different languages, rewrite our concepts, re-write server management docs, and prepare for our hosted cloud offering that will enable all developers to build invincible applications.

Reviewer: ```suggestion
&nbsp;&nbsp; We are far from satisfied with our documentation. We plan to bring SDK reference right into the docs, merge SDK tutorials into a single journey, so that a user can easily see how things are done across different languages, rewrite our concepts, re-write server management docs, and prepare for our hosted cloud offering that will enable all developers to build invincible applications.
```

I think you can say we still need to do lots of improvements without saying "We are far from satisfied". I also wouldn't publicly disclose our specific plans "We plan to bring SDK reference right into the docs, merge SDK tutorials into a single journey," just in case you realize its not a good idea.
